### PR TITLE
Adds rate limiting for participant file downloads

### PR DESCRIPTION
**Pull request by imranmaj**
[https://sagebionetworks.jira.com/browse/BRIDGE-2941](https://sagebionetworks.jira.com/browse/BRIDGE-2941)

Rate limits by bytes

### CODE REVIEW

File: src/main/java/org/sagebionetworks/bridge/BridgeConstants.java [hunk 1]
```diff
@@ -158,4 +158,15 @@ public class BridgeConstants {
...
+
+    public static final long PARTICIPANT_FILE_RATE_LIMITER_INITIAL_BYTES_PROD = 1_000_000; // 1 MB
+    public static final long PARTICIPANT_FILE_RATE_LIMITER_MAXIMUM_BYTES_PROD = 10_000_000; // 10 MB
+    public static final long PARTICIPANT_FILE_RATE_LIMITER_REFILL_INTERVAL_SECONDS_PROD = 3600; // every hr
+    public static final long PARTICIPANT_FILE_RATE_LIMITER_REFILL_BYTES_PROD = 1_000_000; // 1 MB
+
+    public static final long PARTICIPANT_FILE_RATE_LIMITER_INITIAL_BYTES_TEST = 1_000; // 1 KB
+    public static final long PARTICIPANT_FILE_RATE_LIMITER_MAXIMUM_BYTES_TEST = 1_000; // 1 KB
+    public static final long PARTICIPANT_FILE_RATE_LIMITER_REFILL_INTERVAL_SECONDS_TEST = 5; // every 5 seconds
+    public static final long PARTICIPANT_FILE_RATE_LIMITER_REFILL_BYTES_TEST = 1_000; // 1 MB
```
Comment (zaro0508) [inline src/main/java/org/sagebionetworks/bridge/BridgeConstants.java]: shouldn't this be `1KB`?

File: src/main/java/org/sagebionetworks/bridge/BridgeConstants.java [hunk 2]
```diff
@@ -158,4 +158,15 @@ public class BridgeConstants {
             "102", "203", "556", "692", "790", "821", "823", "830", "831", "878", "879", "884", 
             "890", "893");
 
+    public static final String PARTICIPANT_FILE_RATE_LIMIT_ERROR = "User requested to download too much data";
+
+    public static final long PARTICIPANT_FILE_RATE_LIMITER_INITIAL_BYTES_PROD = 1_000_000; // 1 MB
```
Comment (zaro0508) [inline src/main/java/org/sagebionetworks/bridge/BridgeConstants.java]: whaoo! you can assign `1_000_000` as a `long` type?

File: src/main/java/org/sagebionetworks/bridge/BridgeConstants.java (hunk 1 above)
Comment (imranmaj) [reply src/main/java/org/sagebionetworks/bridge/BridgeConstants.java]: Yes thanks

File: src/main/java/org/sagebionetworks/bridge/BridgeConstants.java (hunk 2 above)
Comment (imranmaj) [reply src/main/java/org/sagebionetworks/bridge/BridgeConstants.java]: Yeah it's nice for readability!

File: src/main/java/org/sagebionetworks/bridge/util/ByteRateLimiter.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+import java.time.Instant;
+
+/**
+ * Rate limiter which limits a resource by bytes. The number of bytes that can
+ * be consumed is tracked and is periodically "refilled."
+ */
+public class ByteRateLimiter {
+    // The maximum number of bytes that can be accumulated.
+    private long maximumTokens;
```
Comment (imranmaj) [inline src/main/java/org/sagebionetworks/bridge/util/ByteRateLimiter.java]: maximumBytes

**Review by jnelsonjava — APPROVED** [review]
Looks good!

File: src/main/java/org/sagebionetworks/bridge/services/ParticipantFileService.java [hunk 4]
```diff
@@ -50,6 +70,7 @@ final void setParticipantFileDao(ParticipantFileDao dao) {
     @Autowired
     final void setConfig(BridgeConfig config) {
         bucketName = config.get(PARTICIPANT_FILE_BUCKET);
+        isProduction = config.isProduction();
```
Comment (imranmaj) [inline src/main/java/org/sagebionetworks/bridge/services/ParticipantFileService.java]: Place in BridgeServer2.conf

File: src/main/java/org/sagebionetworks/bridge/BridgeConstants.java [hunk 5]
```diff
@@ -158,4 +158,15 @@ public class BridgeConstants {
...
+    public static final long PARTICIPANT_FILE_RATE_LIMITER_MAXIMUM_BYTES_PROD = 10_000_000; // 10 MB
+    public static final long PARTICIPANT_FILE_RATE_LIMITER_REFILL_INTERVAL_SECONDS_PROD = 3600; // every hr
+    public static final long PARTICIPANT_FILE_RATE_LIMITER_REFILL_BYTES_PROD = 1_000_000; // 1 MB
+
+    public static final long PARTICIPANT_FILE_RATE_LIMITER_INITIAL_BYTES_TEST = 1_000; // 1 KB
+    public static final long PARTICIPANT_FILE_RATE_LIMITER_MAXIMUM_BYTES_TEST = 1_000; // 1 KB
+    public static final long PARTICIPANT_FILE_RATE_LIMITER_REFILL_INTERVAL_SECONDS_TEST = 5; // every 5 seconds
+    public static final long PARTICIPANT_FILE_RATE_LIMITER_REFILL_BYTES_TEST = 1_000; // 1 KB
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/BridgeConstants.java]: You can replace these with config entries in BridgeServer2.conf. See for example https://github.com/Sage-Bionetworks/BridgeServer2/blob/develop/src/main/resources/BridgeServer2.conf#L20

File: src/main/java/org/sagebionetworks/bridge/services/ParticipantFileService.java [hunk 6]
```diff
@@ -58,42 +79,114 @@ final void setS3client(AmazonS3 s3) {
...
+     * @return the size of the file in bytes (0 if not found)
+     */
+    private long getS3FileSize(ParticipantFile file) {
+        try {
+            return s3Client.getObjectMetadata(bucketName, getFilePath(file)).getContentLength();
+        } catch (AmazonS3Exception e) {
+            // file may not have been uploaded yet
+            if (e.getStatusCode() == HttpStatus.SC_NOT_FOUND) {
+                return 0;
+            }
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/services/ParticipantFileService.java:100]: Add unit test case for when getObjectMetadata throws a 404 and another test case for when it throws a different exception (such as 500 Internal Server Error).

File: src/main/java/org/sagebionetworks/bridge/services/ParticipantFileService.java [hunk 7]
```diff
@@ -58,42 +79,114 @@ final void setS3client(AmazonS3 s3) {
...
 
         if (pageSize < API_MINIMUM_PAGE_SIZE || pageSize > API_MAXIMUM_PAGE_SIZE) {
             throw new BadRequestException(PAGE_SIZE_ERROR);
         }
-        return participantFileDao.getParticipantFiles(userId, offsetKey, pageSize);
+        ForwardCursorPagedResourceList<ParticipantFile> files = participantFileDao.getParticipantFiles(userId,
+                offsetKey, pageSize);
+        if (files == null) {
+            return null;
+        }
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/services/ParticipantFileService.java:134]: Add unit test case for when getParticipantFiles returns null.

File: src/test/java/org/sagebionetworks/bridge/services/ParticipantFileServiceTest.java [hunk 8]
```diff
@@ -97,6 +108,19 @@ public void getParticipantFilesPageSizeTooLarge() {
...
 
+    @Test(expectedExceptions = {LimitExceededException.class})
+    public void getParticipantFilesRateLimited() {
+        when(mockFileDao.getParticipantFiles("userid", null, 100)).thenAnswer(invocation -> {
+            List<ParticipantFile> files = new ArrayList<>();
+            for (int i = 0; i < 11; i++) {
+                files.add(ParticipantFile.create());
+            }
+            return new ForwardCursorPagedResourceList<>(files, null, true);
+        });
```
Comment (DwayneJengSage) [inline src/test/java/org/sagebionetworks/bridge/services/ParticipantFileServiceTest.java]: Minor: You can use thenReturn() instead of thenAnswer() to simply return a previously created ForwardCursorPagedResourceList. This way, you won't need to create a lambda, and the code will be simpler and easier to read.

File: src/main/java/org/sagebionetworks/bridge/util/ByteRateLimiter.java [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+ * Rate limiter which limits a resource by bytes. The number of bytes that can
+ * be consumed is tracked and is periodically "refilled."
+ */
+public class ByteRateLimiter {
+    // The maximum number of bytes that can be accumulated.
+    private long maximumBytes;
+    // The time between byte refills in seconds.
+    private long refillIntervalSeconds;
+    // The number of bytes that is refilled every refillIntervalSeconds.
+    private long refillAmount;
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/util/ByteRateLimiter.java]: I recommend marking the immutable fields as final (eg `private final long maximumBytes`). This tells the compiler that this field can never be change, so someone editing this class in the future won't accidentally try to change the value.

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]
Overall looks good. A few minor comments.

File: src/main/resources/BridgeServer2.conf [hunk 10]
```diff
@@ -198,6 +198,26 @@ query.param.allowlist = type,appId,studyId,IdFilter,assignmentFilter,externalId,
...
+
+# Participant file rate limiting test constants
+# 1 KB
+participant-file.rate-limiter.test.initial-bytes = 1000
+# 1 KB
+participant-file.rate-limiter.test.maximum-bytes = 1000
+# 5 s
+participant-file.rate-limiter.test.refill-interval-seconds = 5
+# 1 KB
+participant-file.rate-limiter.test.refill-bytes = 1000
```
Comment (DwayneJengSage) [inline src/main/resources/BridgeServer2.conf]: You can instead use participant-file.rate-limiter.initial-bytes and prod.participant-file.rate-limiter.initial-bytes. All stages will automatically pick up the base entry, and prod will automatically pick up the prod value. You don't need a hard-coded switch on isProduction.

**Review by DwayneJengSage — CHANGES_REQUESTED** [review]

File: src/main/resources/BridgeServer2.conf (hunk 10 above)
Comment (imranmaj) [reply src/main/resources/BridgeServer2.conf]: Ah I see, thanks, that's cool

**Review by DwayneJengSage — APPROVED** [review]
